### PR TITLE
nxos_gir: fix for timeout issue

### PR DESCRIPTION
**Pull request by nkshrishail**
##### SUMMARY
nxos_gir "system mode maintenance" command needs more than 120 seconds to succeed.
One needs to set persistent_timeout env variables to avoid timeout errors for the above command. However, with the timeouts set in env, only localNxapi was working fine. Ansible HttpApi still hit timeout and bailed out.

HttpApi was not making use of persistent_timeout set in env, it was only considering the module parameter "timeout" to set the timeout on the connection. 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
nxos_gir

### CODE REVIEW

**Review by NilashishC — CHANGES_REQUESTED** [review]
Removing `provider` would fail integration tests coz we're still running against connection local.
https://github.com/ansible/ansible/blob/devel/test/integration/targets/nxos_gir/tasks/nxapi.yaml#L30

**Comment by nkshrishail** [comment]
@NilashishC , right, there is another PR coming in soon which will remove the support for local nxapi plugin and will also remove any references of local and also of provider. This one can be merged after the upcoming PR gets merged.

**Comment by mikewiebe** [comment]
> @NilashishC , right, there is another PR coming in soon which will remove the support for local nxapi plugin and will also remove any references of local and also of provider. This one can be merged after the upcoming PR gets merged.

@nkshrishail We discussed on the call today that we can't remove the code in module_utils for nxapi local yet.  Even though it's been deprecated in the docs it has not officially been deprecated in the code.  We can still remove the tests for nxapi local but not the class.  For this PR you just need to remove the following block in addition to removing the `provider` references so that the tests don't run against nxapi local.

https://github.com/ansible/ansible/blob/devel/test/integration/targets/nxos_gir/tasks/nxapi.yaml#L29

Lines 29-33

**Comment by nkshrishail** [comment]
@mikewiebe , ok, the other PR with references to provider and local removed went up just a little while ago, it has the class removed as well, will restore it and update the PR.

**Comment by mikewiebe** [comment]
> @mikewiebe , ok, the other PR with references to provider and local removed went up just a little while ago, it has the class removed as well, will restore it and update the PR.

Thanks.  We only need to restore the class removal.  The tests are fine to remove but it's good to have this reference PR so that when we need to remove the class we have a starting point.

File: lib/ansible/plugins/connection/httpapi.py
Comment (mikewiebe) [inline lib/ansible/plugins/connection/httpapi.py]: @nkshrishail We discussed this PR on the call today.  You should be able to remove the `t1`, `t2` and `timeout` references and just add the following line.

```python
timeout=self.get_option('persistent_command_timeout'),
validate_certs=self.get_option('validate_certs'),
...
```

There a couple of different ways to set this timeout that we need to verify.

1. Set the environment variable: `export ANSIBLE_PERSISTENT_COMMAND_TIMEOUT=600`
2. Set it in the `ansible.cfg` file. https://docs.ansible.com/ansible/latest/reference_appendices/config.html#persistent-command-timeout
3. Passed in as a per/task variable. (Example)

```
- name: save running-config
  ios_command:
    commands: copy running-config startup-config
  vars:
    ansible_command_timeout: 60
```

File: test/integration/targets/nxos_gir/tests/common/sanity.yaml
Comment (mikewiebe) [inline test/integration/targets/nxos_gir/tests/common/sanity.yaml:86]: Let's add one more set of steps to the cleanup section in this test.  I have seen cases where a switch gets stuck in maintenance mode and simply issuing the `no system mode maintenance` command will not work because the command is not even recognized as a valid command.  The following sequence seems to always work (at least in my tests)

1. `configure maintenance profile normal-mode`
2. `no system mode maintenance`

Comment (nkshrishail) [reply test/integration/targets/nxos_gir/tests/common/sanity.yaml:86]: This one may not work in failure scenarios, the command "system mode maintenance" takes 120+ seconds to get into maintenance mode in working cases. If you hit timeout or an error with "system mode maintenance" command and it has not been 120+ seconds, the switch would not let you take it out of maintenance mode. We may have to add bit of delay before trying to clean up in failure scenarios.

File: lib/ansible/plugins/connection/httpapi.py
Comment (nkshrishail) [reply lib/ansible/plugins/connection/httpapi.py]: @mikewiebe all three options work fine.

File: test/integration/targets/nxos_gir/tests/common/sanity.yaml
Comment (mikewiebe) [reply test/integration/targets/nxos_gir/tests/common/sanity.yaml:86]: Good points.  I am starting to lean towards adding code to make the module itself more robust.  We should update the action plugin for nxos set the timeout automatically for this module instead of forcing the user to set a timeout that is large enough.  Here is an example of how I am warning users to set the timeout to a value large enough when upgrading images using the `nxos_install_os` module.

https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/nxos.py#L67

In this case for the  `nxos_gir` module we could just set it to say 200 seconds to allow enough time for the maintenance mode to complete and then add logic in the module to handle any timeout that is beyond 200 seconds to try and roll back to the previous state.

File: lib/ansible/module_utils/network/nxos/nxos.py
Comment (NilashishC) [inline lib/ansible/module_utils/network/nxos/nxos.py:628]: This should also check for `provider['timeout']` value. Something like `persistent_command_timeout = C.PERSISTENT_COMMAND_TIMEOUT or self._module.get(‘provider’, {}).get(‘timeout’)`.

Comment (chrisvanheuveln) [reply lib/ansible/module_utils/network/nxos/nxos.py:628]: Good catch. I'll add a check.

edit: Nevermind. Found that this check isn't necessary since `connection.transport` will never equal `'local'`, as this check lives in `class Httpapi`. I simplified the check in the rewrite.

**Comment by chrisvanheuveln** [comment]
@nkshrishail @mikewiebe @NilashishC I'm taking over this PR from Shrishail. I've addressed Mike and Nilashish comments. 
Unfortunately I can't commit to his repo so please close this one. The new PR is #66043 .

**Comment by NilashishC** [comment]
Replaced by #66043

**Pull request closed without merging**
